Sources/SourceKitLSP/Swift/Refactoring.swift

```diff
@@ -0,0 +1,130 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LSPLogging
+import LanguageServerProtocol
+import SourceKitD
+
+protocol RefactoringResponse {
+  init(title: String, uri: DocumentURI, refactoringEdits: [RefactoringEdit])
+}
+
+extension RefactoringResponse {
+  /// Create an instance of `RefactoringResponse` from a sourcekitd semantic
+  /// refactoring response dictionary, if possible.
+  ///
+  /// - Parameters:
+  ///   - title: The title of the refactoring action.
+  ///   - dict: Response dictionary to extract information from.
+  ///   - snapshot: The snapshot that triggered the `semantic_refactoring` request.
+  ///   - keys: The sourcekitd key set to use for looking up into `dict`.
+  init?(_ title: String, _ dict: SKDResponseDictionary, _ snapshot: DocumentSnapshot, _ keys: sourcekitd_api_keys) {
+    guard let categorizedEdits: SKDResponseArray = dict[keys.categorizedEdits] else {
+      logger.error("categorizedEdits doesn't exist in response dictionary")
+      return nil
+    }
+
+    var refactoringEdits = [RefactoringEdit]()
+
+    categorizedEdits.forEach { _, categorizedEdit in
+      guard let edits: SKDResponseArray = categorizedEdit[keys.edits] else {
+        logger.error("edits doesn't exist in categorizedEdit dictionary")
+        return true
+      }
+      edits.forEach { _, edit in
+        // The LSP is zero based, but semantic_refactoring is one based.
```

Comment:
I think this comment makes more sense before the lines where you do the `- 1` calculations.